rdm/init_files/documents/verification_and_validation_plan.md

```diff
@@ -0,0 +1,78 @@
+
+A **unit test** examines an individual software unit that typically cannot be broken down any further.
+
+An **integration test** examines multiple software units for defects in the interfaces and interactions between them.
+
+An **automated test** is any test that can be run without human input. A **manual test** is any test that requires a human to perform some or all of the test steps.
+
+A **software item** is any identifiable part of a computer program.
+
+A **software system** is an integrated collection of software items organized to accomplish a specific function or set of functions.
+
+A **software unit** is a software item that is not subdivided into other software items.
+
+**Regression testing** is the process of re-running functional and non-functional tests to ensure that previously developed and tested software still performs after a change.
+
+# Verification Activities
+
+Software verification begins with the code reviews and adherence to software standards defined in the Software Plan. The outputs of these activities are recorded in the Revision Level History.
+
+Software verification continues by running all automated tests on a continuous basis during the development process to detect any regressions. Automated tests are automatically run against each proposed software revision before the revision is accepted, a process known as Continuous Integration, or CI. Manual tests are run as needed during development and then for the Software Test Record for a release.
+
+Tests are written, modified, and reviewed during the Unit Implementation and Testing activity and are reviewed for completeness as part of the Release activity. Both activities are described in the Software Plan.
+
+A detailed list of verification tests, the testing environment, and their results are found in the Software Test Record. The Software Test Record is created as part of the Integration and System Testing activity described in the Software Plan.
+
+The types of tests used to verify {{ device.name }} are described below:
+
+## Unit Tests
+
+TODO: Describe how and when unit test shall be written
+
+## Integration Tests
+
+TODO: Describe how and when integration tests shall be written
+
+## Automated End-to-End Tests
+
+TODO: Describe how and when automated end-to-end tests shall be written
+
+## Manual End-to-End Tests
+
+TODO: Describe how and when manual end-to-end tests shall be written
+
+# Validation Activities
+
+System-level validation of {{ device.name }} is covered by human factors and clinical testing. Thus, it will not be covered in
+this software-specific documentation.
+
+TODO: Reference the appropriate documents
+
+# Pass / Fail Criteria
+
+TODO:
+
+This plan should include a pass/fail criteria for the entire test suite.  E.g., you require that all unit tests pass and that all integration tests pass or the cause of the failure is understood and justified
+
+ENDTODO
+
+[[62304:5.7.1.a]]
```

Comment:
This looks great. One comment the regulatory consultant for a client brought up is that it is a little confusing that we only list the actual tests that are executed and each individual manual test step in the test record (the document signed after all tests are executed). He was under the impression we should have these all enumerated as a part of the testing plan, and I think that makes sense-- we want the tests being performed to be signed off before we perform the tests and record their results.

What do you think about this? I think practically, this would look like a new section at the bottom of the V&V plan that is really similar to the test record document, but just enumerates all the automated tests and lists out all the manual test steps.

Reply:
I agree that it makes sense to review all of the tests prior to running them, but I'm not sure I think it's good to have a plan document be updated each time new tests are added. With the way that the current RDM plan is set up, the test coverage is reviewed as part of the release checklist (see https://github.com/innolitics/rdm/blob/main/rdm/init_files/documents/release_record.md#release-verification), although it doesn't really explain _how_ you'd do this review. It seems like we would want to have the list available somewhere and for the person reviewing it to have a record of what they signed off on... it seems fine to me that the test record implicitly contains this.

Thus, it could work like this:

1. Tech lead begins working through the release checklist
2. They read through all of the tests that are available, perhaps just looking at the generated traceability matrix and the empty test record
3. If it looks good, they kick off the testing activity
4. A test record is produced
5. The release record references the test record they signed off on (which, presumably would need to match what they were expecting from step 2)

Reply:
@johndgiese I think this makes sense to me, but I want to make sure I understand:

> I'm not sure I think it's good to have a plan document be updated each time new tests are added.

It seems like the test plan, as currently written, describes how we will verify the device and what our acceptance criteria are for passing at a high level. e.g. we say that all unit tests need to pass and all integration tests need to pass or have a justification for failure. However, we don't actually define what the specifics of those tests are, or even how many of them there are. It's more just describing the approach we take for V&V as a process.

I've seen some clients refer to a test _protocol_, which basically just looks like the test record without the place to record results. I previously thought this was the same thing as the test plan, but now I am starting to think these are distinct concepts. If I understand your point, you are saying that it doesn't make sense to keep these specifics in the plan, and the test record can serve both as a record of results and a description of the test steps. Signing off on the test record would then mean that it communicates two things:

1. The test results in the record are accurate
2. The tests and test steps described in the document are approved

Is that an accurate summary of our approach here?

Reply:
Sorry this got so long---a lot of this is my notes as I thought about the problem. Don't feel like you need to read it all carefully.

# Intro

> It seems like the test plan, as currently written, describes how we will verify the device and what our acceptance criteria are for passing at a high level. e.g. we say that all unit tests need to pass and all integration tests need to pass or have a justification for failure. However, we don't actually define what the specifics of those tests are, or even how many of them there are. It's more just describing the approach we take for V&V as a process.

This is all consistent with what I'd think the plan should contain.

> I've seen some clients refer to a test protocol, which basically just looks like the test record without the place to record results. I previously thought this was the same thing as the test plan, but now I am starting to think these are distinct concepts.

I'm less familiar with the term protocol, so I'd be curious to know how the client was using that term. I know that it is used and have seen it a few times, so I'm sure there is a meaning. I am familiar with the terms policies, processes, plans, procedures, and practices. I'm trying to figure out how people usually use these words so we can communicate better with our clients. Then, I'm also wondering what distinctions are useful for RDM (I'm sure all of the distinctions are useful sometimes to some people).

# How people usually use the terms

I looked in ISO 13485 and ISO 9000, and neither use the term protocol. 21 CFR 820 only uses the protocol once in 820.70(i)---the tools validation part of the regulation:

> (i) Automated processes. When computers or automated data processing systems are used as part of production or the quality system, the manufacturer shall validate computer software for its intended use according to an established protocol. All software changes shall be validated before approval and issuance. These validation activities and results shall be documented.

There is no definition of the term here. The term _procedure_ and _process_ are used many times (e.g., > 30) in the regulation.

According to ISO 9000, a _procedure_ is a specified way to carry out an activity or a process. A _process_ is a set of interrelated activities that use inputs to deliver outputs. An _activity_ is a smaller chunk of work that takes inputs and produces outputs. A _policy_ is the intentions and direction of an organization as formally expressed by its top management.

As best I can tell, a plan can mean a few different things depending on the situation.

Here's where the term _plan_ is used a few times in the regulation:

> Quality planning. Each manufacturer shall establish a quality plan which defines the quality practices, resources, and activities relevant to devices that are designed and manufactured. The manufacturer shall establish how the requirements for quality will be met.

> Design and development planning. Each manufacturer shall establish and maintain plans that describe or reference the design and development activities and define responsibility for implementation. The plans shall identify and describe the interfaces with different groups or activities that provide, or result in, input to the design and development process. The plans shall be reviewed, updated, and approved as design and development evolves.

> Sampling plans, when used, shall be written and based on a valid statistical rationale. Each manufacturer shall establish and maintain procedures to ensure that sampling methods are adequate for their intended use and to ensure that when changes occur the sampling plans are reviewed. These activities shall be documented.

 and a bunch of times in ISO 13485. In the first screenshot, they mention verification plans in particular:

![Screen Shot 2021-12-13 at 9 33 41 AM](https://user-images.githubusercontent.com/1648754/145831262-c2e3b7c7-efd5-480b-a2ea-e9dbac702718.png)

Here are some other plans they mention, which mostly parallel what is in 21 CFR 820:

![Screen Shot 2021-12-13 at 9 33 01 AM](https://user-images.githubusercontent.com/1648754/145831265-bb851e42-02fb-41ea-b629-4c75db55ee26.png)
![Screen Shot 2021-12-13 at 9 32 40 AM](https://user-images.githubusercontent.com/1648754/145831266-69e07a47-5ebb-4f7c-a128-e543181e6801.png)
![Screen Shot 2021-12-13 at 9 32 10 AM](https://user-images.githubusercontent.com/1648754/145831267-b9eb34bb-ff60-4288-895f-cc0120006f3d.png)

So, after reading through all of this, I'm still not clear on what people usually mean by a _plan_. I think I know what a _procedure_, _process_, and _policy_ mean. I'm also not sure what a _protocol_ is.

# What is useful?

Now to the second question: What distinctions are useful for us? I'm not exactly sure. It seems to me like here are the requirements we want to meet:

1. During development, engineers can write automated or manual test steps (or other verification activities) more or less as they construct the code.
2. We have a verification record of what tests and other verification activities were actually done.
3. The person who did the verification signs off on this record.
4. For each release, someone (probably the tech lead) verifies the verification procedure.
5. There is a record of this person signing off on that.
6. This record ideally would be created before verification starts, although what mostly matters is that the verification procedure be verified before it's ran.
7. Our approach reduces duplication of information
8. Our approach avoids unnecessary document changes

What do you think about these, and (6) in particular?

> Signing off on the test record would then mean that it communicates two things:
> 
> 1. The test results in the record are accurate
> 2. The tests and test steps described in the document are approved
> 
> Is that an accurate summary of our approach here?

This is a bit different than what I was thinking. In particular, I think the person signing off that the verification procedure is good won't be the same person who runs the procedure. E.g., I bet the tech lead will often review the procedure but a manual QA person will do the manual test procedure. The manual QA person won't be able to sign off that the verification procedure is sufficient.

Thus, I was proposing something a bit different. The test record signoff just indicates that the steps in the tests were done. The release record signoff indicates a bunch of things, including that the test coverage was sufficient.

If the test coverage is not sufficient, then the tech lead may need to ask the tester to run through another set of tests. In general though, the tech lead will review the test coverage before asking the manual QA person to run through them, even if they sign off on the release record after they're done.

Reply:
@ReeceStevens we should continue this conversation, but I'm going to go ahead and merge the PR.